components/eventing-controller/controllers/subscription/nats/reconciler.go

```diff
@@ -49,43 +51,65 @@ type Reconciler struct {
+	customEventsChannel chan event.GenericEvent
 }
 
 var (
 	Finalizer = eventingv1alpha1.GroupVersion.Group
 )
 
 const (
-	NATSProtocol          = "NATS"
 	NATSFirstInstanceName = "eventing-nats-1" // NATSFirstInstanceName the name of first instance of NATS cluster
 	NATSNamespace         = "kyma-system"     // NATSNamespace namespace of NATS cluster
 	reconcilerName        = "nats-subscription-reconciler"
 	clusterLocalURLSuffix = "svc.cluster.local"
 )
 
-func NewReconciler(ctx context.Context, client client.Client, applicationLister *application.Lister, cache cache.Cache, logger *logger.Logger, recorder record.EventRecorder, cfg env.NatsConfig, subsCfg env.DefaultSubscriptionConfig) *Reconciler {
-	natsHandler := handlers.NewNats(cfg, subsCfg, logger)
+func NewReconciler(ctx context.Context, client client.Client, applicationLister *application.Lister, cache cache.Cache,
+	logger *logger.Logger, recorder record.EventRecorder, cfg env.NatsConfig, subsCfg env.DefaultSubscriptionConfig) *Reconciler {
+	reconciler := &Reconciler{
+		ctx:                 ctx,
+		Client:              client,
+		Cache:               cache,
+		logger:              logger,
+		recorder:            recorder,
+		eventTypeCleaner:    eventtype.NewCleaner(cfg.EventTypePrefix, applicationLister, logger),
+		sinkValidator:       defaultSinkValidator,
+		customEventsChannel: make(chan event.GenericEvent),
+	}
+	natsHandler := handlers.NewNats(cfg, subsCfg, reconciler.handleNatsConnClose, logger)
 	if err := natsHandler.Initialize(env.Config{}); err != nil {
 		logger.WithContext().Errorw("start reconciler failed", "name", reconcilerName, "error", err)
 		panic(err)
 	}
+	reconciler.Backend = natsHandler
+
+	return reconciler
+}
 
-	return &Reconciler{
-		ctx:              ctx,
-		Client:           client,
-		Cache:            cache,
-		Backend:          natsHandler,
-		logger:           logger,
-		recorder:         recorder,
-		eventTypeCleaner: eventtype.NewCleaner(cfg.EventTypePrefix, applicationLister, logger),
-		sinkValidator:    defaultSinkValidator,
+// handleNatsConnClose is called by NATS when the connection to the NATS server is closed. When it
+// is called, the reconnect attempts have exceeded the defined value.
+// It force reconciling the subscription to make sure the subscription is marked as not ready, until
+// it is possible to connect to the NATS server again.
+// See https://github.com/kyma-project/kyma/issues/12930
+func (r *Reconciler) handleNatsConnClose(_ *nats.Conn) {
+	r.namedLogger().Info("NATS connection is closed and reconnect attempts are exceeded")
+	subs, err := r.getAllKymaSubscriptions(context.Background())
+	if err != nil {
+		// NATS reconnect attempts are exceeded, and we cannot reconcile subscriptions! If we ignore this,
+		// there will be no future chance to retry connecting to NATS!
+		panic(err)
 	}
+	r.enqueueReconciliationForSubscriptions(subs)
 }
 
-func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
```

Comment:
Is there any reason to remove this method? Isn't it currently referenced in [here](https://github.com/kyma-project/kyma/blob/9aaa0ee225fa06d62b5c58ce17093156c15374fe/components/central-application-connectivity-validator/cmd/centralapplicationconnectivityvalidator/centralapplicationconnectivityvalidator.go#L136)?

Reply:
I hope not! :-) Otherwise, it shouldn't build! This was not used. I think it is just left over from before, when the subscription controller were used separately. This is int he NATS reconciler file. What you referenced is returned from a `NewController` call, I think.